src/Optimizer/HeroCandidateFiltering.php

```diff
@@ -0,0 +1,131 @@
+<?php
+/**
+ * Class HeroCandidateFiltering.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP\Optimizer;
+
+use AmpProject\AmpWP\Infrastructure\Conditional;
+use AmpProject\AmpWP\Infrastructure\Delayed;
+use AmpProject\AmpWP\Infrastructure\Registerable;
+use AmpProject\AmpWP\Infrastructure\Service;
+use AmpProject\Attribute;
+use WP_Post;
+
+/**
+ * Service which uses WordPress hooks to inject data-hero-candidate attributes on logos, headers, and the primary featured image.
+ *
+ * @package AmpProject\AmpWP
+ * @since 2.1.0
+ * @internal
+ */
+final class HeroCandidateFiltering implements Service, Delayed, Conditional, Registerable {
+
+	/**
+	 * Whether the Custom Logo has been filtered.
+	 *
+	 * This is set to true after the first filtering to prevent filtering the logo in the footer, for example.
+	 *
+	 * @var bool
+	 */
+	protected $custom_logo_filtered = false;
+
+	/**
+	 * Get the action to use for registering the service.
+	 *
+	 * @return string Registration action to use.
+	 */
+	public static function get_registration_action() {
+		return 'wp';
+	}
+
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed() {
+		return amp_is_request();
```

Comment:
I'll have to pay attention to this one. It means we need to ensure in the CLI context that we need to ensure this evaluates to true, otherwise these filters will be missing.

Reply:
Good point. It will always be false in a WP-CLI context currently:

https://github.com/ampproject/amp-wp/blob/1e7ea02cee63910aa5fe20c38e9f588cb6bdb170/includes/amp-helper-functions.php#L327-L329